### ISSUE TITLE
Conv2D with `same` padding + asymmetric stride returns all-zeros in the LiteRT interpreter

### ISSUE THREAD

**Issue by ali-nrn** [google-ai-edge/LiteRT#9256]
# Conv2D with `same` padding + asymmetric stride returns all-zeros in the LiteRT interpreter

## Summary

For a `Conv2D` with `padding="same"` and certain **asymmetric strides** (e.g. `(3, 4)` or `(4, 3)`), the `ai_edge_litert` interpreter returns an
output that is **entirely zeros**, while the same model evaluated in Keras/TF produces the correct nonzero result. The exported `.tflite` model appears to be correct but the interpreter's numeric output is wrong.

The failure is:
- **silent** — no error or warning is raised; the interpreter just returns zeros;
- **deterministic** — reproduces on every run with a fixed seed;
- **independent of data format** — reproduces in plain `channels_last` (NHWC) as  well as `channels_first` (NCHW);
- **specific to `same` padding** — the identical model with `padding="valid"` is  always correct;
- **specific to certain asymmetric strides** — symmetric strides are always correct.

## Environment

| Package | Version |
|---|---|
| `ai-edge-litert` | **2.1.5** |
| `tensorflow` | 2.18.0 |
| `keras` | 3.12.0 |
| `numpy` | 2.0.2 |
| Python | 3.12 |
| Platform | Linux x86-64, CPU (no delegate; default `BUILTIN` op resolver) |

## Minimal reproducer

Full script attached: **`repro_litert_asymmetric_stride_same.py`**.
Minimal essence (plain NHWC, batch 1, bias, no transposes):

```python
import numpy as np, tensorflow as tf, keras
from ai_edge_litert.interpreter import Interpreter

np.random.seed(0)
x = np.random.uniform(-1, 1, size=(1, 8, 8, 16)).astype(np.float32)

inp = keras.layers.Input(batch_shape=(1, 8, 8, 16))
out = keras.layers.Conv2D(8, 3, strides=(3, 4), padding="same",
                          data_format="channels_last", use_bias=True)(inp)
model = keras.models.Model(inp, out)

y_keras = model.predict(x, verbose=0)                     # correct, nonzero

tflite = tf.lite.TFLiteConverter.from_keras_model(model).convert()
it = Interpreter(model_content=tflite); it.allocate_tensors()
di, do = it.get_input_details()[0], it.get_output_details()[0]
it.set_tensor(di["index"], x.astype(di["dtype"])); it.invoke()
y_litert = it.get_tensor(do["index"])                     # ALL ZEROS

print("keras  nonzero:", np.any(y_keras != 0))            # True
print("litert all-zero:", np.all(y_litert == 0))          # True
print("max abs diff:", np.max(np.abs(y_keras - y_litert)))# ~2.1
```

## Observed output

```
[1] MINIMAL CASE  (NHWC, input 1x8x8x16, Conv2D 8 filters, kernel 3x3, strides=(3,4), padding='same')
    keras : shape=(1, 3, 2, 8) zeros=0/48  mean|.|=0.5567
    litert: shape=(1, 3, 2, 8) zeros=48/48 mean|.|=0
    max_abs_diff = 2.11161  --> BUG REPRODUCED (litert all-zeros)
```

Note the shapes match — the interpreter produces the correct **shape**, just
all-zero **values**.

## What triggers it (trigger table)

NHWC, input `8x8x16`, kernel `3x3`, with bias. `max_abs_diff` is between Keras and the interpreter:

| strides | padding | max_abs_diff | result |
|---|---|---|---|
| (1, 1) | valid | 1e-06 | ok |
| (1, 1) | same  | 7e-07 | ok |
| (2, 2) | valid | 5e-07 | ok |
| (2, 2) | same  | 5e-07 | ok |
| (3, 3) | valid | 4e-07 | ok |
| (3, 3) | same  | 5e-07 | ok |
| (2, 1) | valid | 1e-06 | ok |
| (2, 1) | same  | 1e-06 | ok |
| (1, 2) | valid | 8e-07 | ok |
| (1, 2) | same  | 7e-07 | ok |
| (2, 3) | valid | 4e-07 | ok |
| (2, 3) | same  | 3e-07 | ok |
| (3, 4) | valid | 2e-07 | ok |
| **(3, 4)** | **same** | **1.675** | **>>> BUG (all-zeros)** |
| (4, 3) | valid | 4e-07 | ok |
| **(4, 3)** | **same** | **1.261** | **>>> BUG (all-zeros)** |

## Likely root-cause hint

The failing strides are exactly those where `same` padding resolves to a **nonzero pad in one spatial dimension and zero pad in the other**:

For input size 8, kernel 3:
- stride 3 → output `ceil(8/3)=3`, required pad `= (3-1)*3 + 3 - 8 = 1`
- stride 4 → output `ceil(8/4)=2`, required pad `= (2-1)*4 + 3 - 8 = -1 → 0`

So `(3, 4)` needs pad `(1, 0)` and `(4, 3)` needs pad `(0, 1)` — one axis padded, the other not. By contrast `(2, 3)` needs pad `(1, 1)` (both axes padded) and works correctly. This points at the handling of the implicit `SAME` padding `PAD`-fusion when the computed padding is zero along one axis but not the other.

See repro script:
[repro_litert_asymmetric_stride_same.py](https://github.com/user-attachments/files/30991446/repro_litert_asymmetric_stride_same.py)

**Comment by shivvaam0001** [comment]
Hi @ali-nrn,

Thank you for filing this issue.
To investigate this issue further, could you please provide a few more details?

* Could you please attach the generated `.tflite` model file directly to this issue?
* Could you confirm whether this issue is strictly limited to `float32` inputs and weights, or whether the same zero output behavior occurs with other precision formats as well?


Thanks!

**Comment by jl45G** [comment]
Hey Terry, could you help take a look?

**Comment by ali-nrn** [comment]
@shivvaam0001 The script in the description creates a tflite file. With further investigation Claude thinks for a `Conv2D` with `padding="same"` and certain asymmetric strides (e.g. `(3, 4)` or `(4, 3)`), the `ai_edge_litert` interpreter discards the convolution contribution and returns only the bias.

This seems to also apply to other precisions such as float16, and int8. I tried to attach the tflite file but github does not allow.

[repro_litert_asymmetric_stride_same.py](https://github.com/user-attachments/files/31266063/repro_litert_asymmetric_stride_same.py)

**Comment by aravindmurali22** [comment]
Hi @terryheo , following up on this thread, is there any recent updates on this? Thanks!
